californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -508,11 +529,12 @@ private KeyToken createUnusedToken() {
 				KeyMID idByMID = new KeyMID(exchange.getCurrentRequest().getMID());
 				KeyToken idByToken = new KeyToken(exchange.getCurrentRequest().getToken());
 
-				exchangesByToken.remove(idByToken);
-
-				// in case an empty ACK was lost
-				exchangesByMID.remove(idByMID);
+				if (!exchange.getRequest().isMulticast() || isMulticastTimedOut(exchange)) {
+					exchangesByToken.remove(idByToken);
```

Comment:
Instead of using the reception of a response as the trigger to finally remove the exchange from the token map, I think we should instead schedule a task for cleaning up the multicast exchange after 

`NON_LIFETIME + MAX_LATENCY + MAX_SERVER_RESPONSE_DELAY`

as indicated by RFC 7390. This clean-up task would need to be scheduled when the original multicast request is sent out.